### ISSUE TITLE
Ambiguous type name warning for non-ambiguous type name in OracleDatabase

### ISSUE THREAD

**Issue by mdaepp** [jOOQ/jOOQ#15096]
### Expected behavior

Class TPermissionsTab is generated.

### Actual behavior

Class TPermissionsTab is not generated.

```
[WARNING] Ambiguous type name      : The database object SAKILA.CUSTOMERS.T_PERMISSIONS_TAB generates a class org.jooq.sakila.packages.customers.udt.TPermissionsTab (class org.jooq.meta.oracle.OracleUDTDefinition) which conflicts with the previously generated class org.jooq.sakila.packages.customers.udt.TPermissionsTab (class org.jooq.meta.DefaultArrayDefinition). Use a custom generator strategy to disambiguate the types. More information here:
 - https://www.jooq.org/doc/latest/manual/code-generation/codegen-generatorstrategy/
 - https://www.jooq.org/doc/latest/manual/code-generation/codegen-matcherstrategy/
```

### Steps to reproduce the problem

1. Use https://github.com/jOOQ/sakila/tree/main/oracle-sakila-db as a base.
2. Modify oracle-sakila-schema-pl-sql.sql as follows:

```sql
CREATE PACKAGE customers AS
  TYPE person IS RECORD (
    first_name VARCHAR2(50),
    last_name VARCHAR2(50)
  );
  -- This type:
  TYPE t_permissions_tab IS TABLE OF NUMBER(3) INDEX BY VARCHAR2(30 CHAR);
  FUNCTION get_customer(p_customer_id NUMBER) RETURN person;
END customers;
/
```

3. Generate code

Config:

```xml
<generate>
  <validationAnnotations>true</validationAnnotations>
</generate>
```

### jOOQ Version

jOOQ 3.18.4 trial edition

### Database product and version

Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production

### Java Version

openjdk 17.0.7 2023-04-18

### OS Version

Microsoft Windows [Version 10.0.19045.2965]

### JDBC driver name and version (include name if unofficial driver)

oracle.jdbc.OracleDriver 19.3.0.0.0

**Comment by lukaseder** [comment]
Thanks for your report. I think the warning is cosmetic, at least on my attempt to reproduce this. I cannot reproduce this:

> Class TPermissionsTab is not generated.

I'm getting this class here:

```java
/*
 * This file is generated by jOOQ.
 */
package org.jooq.test.oracle.generatedclasses.test.packages.pkg_15097.udt.records;


import org.jooq.impl.AssociativeArrayRecordImpl;
import org.jooq.impl.SQLDataType;
import org.jooq.test.oracle.generatedclasses.test.Test;
import org.jooq.test.oracle.generatedclasses.test.packages.Pkg_15097;


/**
 * This class is generated by jOOQ.
 */
@SuppressWarnings({ "all", "unchecked", "rawtypes" })
public class TPermissionsTabRecord extends AssociativeArrayRecordImpl<String, Short> {

    private static final long serialVersionUID = 1L;

    /**
     * Create a new <code>TEST.PKG_15097.T_PERMISSIONS_TAB</code> record
     */
    public TPermissionsTabRecord() {
        super(Test.TEST, Pkg_15097.PKG_15097, "T_PERMISSIONS_TAB", SQLDataType.SMALLINT, SQLDataType.VARCHAR);
    }
}
```

Is there anything else in your code generation configuration that might help reproduce the case where the class isn't generated?

**Comment by lukaseder** [comment]
The problem is the query that fetches UDTs from Oracle 18c's new `ALL_PLSQL_TYPES` view (see #7861). It is currently written as:

```sql
  ..
  union
  select
    p.OWNER OWNER,
    p.PACKAGE_NAME pkg,
    p.TYPE_NAME TYPE_NAME,
    0 synthetic,
  from SYS.ALL_PLSQL_TYPES p
  where p.OWNER in (
    'TEST'
  )
  union
  ..
```

An additional predicate is needed:

```sql
  and p.TYPECODE != 'COLLECTION'
```

Both associative arrays and other `PL/SQL TABLE` types should be reported as `Database::getArrays`, not `Database::getUDTs`.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.19.0, 3.18.5 (https://github.com/jOOQ/jOOQ/issues/15111), 3.17.14 (https://github.com/jOOQ/jOOQ/issues/15112), 3.16.20 (https://github.com/jOOQ/jOOQ/issues/15113).

Again, this addresses the wrong warning that is being logged, not any potential omission of generated files, which I could not yet reproduce.

**Comment by mdaepp** [comment]
Thanks for the fix! You are right, I was thrown off by the misleading warning that said that a TPermissionsTab class has been generated. I agree that no such class is needed. TPermissionsTabRecord is generated in my case as well, so no worries there.
